### PR TITLE
Docsp 9568 compound

### PR DESCRIPTION
**Pull request by biniona-mongodb**
## Pull Request Info
Adding Fundamentals > CRUD > Compound Ops

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-9568

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/61927d1/java/docsworker-xlarge/DOCSP-9568-Compound/fundamentals/crud/compound-operations/

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- Links to Github source code are expected to not work until this PR is merged into master. Aside from these, all links should be functional.
- [x] Are the staging links in the PR description updated?

### CODE REVIEW

File: source/fundamentals/crud/compound-operations.txt [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+  instance. If the operation takes longer, the ``findOneAndUpdate()`` method
+  will throw a ``MongoExecutionTimeoutException``.  
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java
+   :language: java
+   :dedent:
+   :start-after: start findOneAndUpdate-example
+   :end-before: end findOneAndUpdate-example
+
+The output of the above code should look like this:
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:85]: Suggestion: I think it would be beneficial to explain to the reader what your code is printing.

Applies to all instances.

**Review by kyuan-mongodb — CHANGES_REQUESTED** [review]
Overall, I think the page is pretty good!  I really liked how you engage with the reader in your Race Condition section. 

Most of my comments are around my suggestions in reformatting and clarifying some terms.

File: source/fundamentals/crud/compound-operations.txt [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+On this page you will learn how to perform **compound operations** with the
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:16]: Sounds good, I'll make the change. 

Might be beneficial to discuss this in team PR review and see if this is something we want to include in the template.

File: source/fundamentals/crud/compound-operations.txt [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+MongoDB Java driver. 
+
+Compound operations consist of a read and write operation performed as one
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:19]: Thanks for the note. 

If it is alright, and there are no guidelines I am violating, I like my current phrasing so I would prefer to not make the change.

Note on how I took this piece of feedback: As this had a "Nit" heading, I felt it was appropriate for me to make a judgment call on which I preferred without much discussion or supporting evidence (I read both aloud and picked which one I preferred). If you would like further discussion on this issue, please let me know in this thread or over slack or zoom and I would be down to discuss more.

File: source/fundamentals/crud/compound-operations.txt [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+**atomic operation**. An atomic operation is an irreducible series of one or
+more database operations such that all occur or none occur.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:21]: Thanks for the note. I believe the proposed definition involving "interruptions" is slightly incorrect, as I believe atomic operations still undergo context switching on the OS (and thus are periodically interrupted). I don't know how to defend this claim with primary evidence without digging into the MongoDB server codebase, but in the implementation section of the Wikipedia article on [Database Atomicity](https://en.wikipedia.org/wiki/Atomicity_(database_systems)#Implementation) it states that "Portable operating systems cannot simply block interrupts to implement synchronization, since hardware that lacks concurrent execution such as hyper-threading or multi-processing is now extremely rare." In this quote, I am assuming that portable is used as in Portable Operating System Interface (POSIX).

Because of the above argument, I believe the following definition provided by Jeff Allen is more accurate (there is now a ticket on server for adding this definition to the glossary).

> An atomic operation is an operation which either completes entirely, or does not complete at all. Atomic operations cannot partially complete.

Update to new definition:

File: source/fundamentals/crud/compound-operations.txt [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+- Find and update one document
+- Find and replace one document
+- Find and delete one document
+
+If you need to perform more complex tasks atomically, such as reading and
+writing to more than one document, use **transactions**.
+
+Transactions are a feature of MongoDB and other databases that lets you define
+an arbitrary sequence of database commands as an atomic operation.  
+
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:38]: Thanks for the note. I agree that these should be combined. I can see how my tendency towards 1 sentence paragraphs is not helpful here. Going forward, I will try to be more critical towards my tendency to make choppy sentences.

Fix:

File: source/fundamentals/crud/compound-operations.txt [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    {"_id": 1, "food": "donut", "color": "green"}
+    {"_id": 2, "food": "pear", "color": "yellow"}
+
+`The full code for the following examples is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java>`__
+
+Find and Update One Document
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To find and update a document, use the ``findOneAndUpdate()`` method of the
+``MongoCollection`` class.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt]: Thanks for the note. I agree it is beneficial to the page to shorten the subheadings.

Fix: https://github.com/mongodb/docs-java/pull/80/commits/413c446483fad4ab96deeb173ffd09c954f30be6

File: source/fundamentals/crud/compound-operations.txt [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+.. code-block:: none
+
+   You got the Blue Room Jan
+
+And Pat sees this output:
+
+.. code-block:: none
+
+   You got the Blue Room Pat
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:212]: Thanks for the note. I would like to keep the guest names for the following reasons:

- As concurrency is something that is often hard for people to understand, I want to try and be as clear as possible when things are happening on different machines, and I think having one message could allow people to misread this and think that only one machine/process is involved. 
- The runnable code sample will be hard to understand without guest names included. Currently the java program can be run, and it will be hard to follow without guest names.

File: source/fundamentals/crud/compound-operations.txt [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+.. code-block:: none
+
+   You got the Blue Room Jan
+
+And Pat sees this output:
+
+.. code-block:: none
+
+   Sorry, we are booked Pat
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:262]: Thanks for the note. I believe this edit introduces a temporal relationship between the two examples which I would rather not include. To me, "Now" suggests that Pat's two experiences with this sample app exist in relation to one another, which I think adds a layer of complexity to the examples that obfuscates the crux of the race condition being discussed.

File: source/fundamentals/crud/compound-operations.txt [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+--------
+
+On this page you will learn how to perform **compound operations** with the
+MongoDB Java driver. 
+
+Compound operations consist of a read and write operation performed as one
+**atomic operation**. An atomic operation is an irreducible series of one or
+more database operations such that all occur or none occur.
+
+Atomic operations can help avoid certain kinds of **race conditions** in your code. A
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt]: Thanks for the note. I agree that "certain kinds" unnecessarily introduces uncertainty into this sentence.

Fix:

File: source/fundamentals/crud/compound-operations.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:62]: Thanks for the note. I agree that including return information could benefit readers. I assumed this was understood, but after rereading I see that it is not explicitly stated and probably should be.

Fix: 

I would not like to include parameter information for this reason:

- The compound operation methods have up to 9 method signatures. I believe that providing an  exhaustive list would obfuscate the conceptual content of the page and duplicate API docs content. I think providing an incomplete list of signatures would give readers a false sense of understanding, so I think not including this information here and allowing readers to consult API documentation makes sense.

File: source/fundamentals/crud/compound-operations.txt [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+------------------------------
+
+This section shows how to use each compound operation with the MongoDB Java Driver.
+
+The following examples use a collection containing these two sample documents.
+
+.. code-block:: json
+
+    {"_id": 1, "food": "donut", "color": "green"}
+    {"_id": 2, "food": "pear", "color": "yellow"}
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:54]: Thanks for the note. The json formatting on the page is meant to mimic the default behavior of the Java Driver's `BsonDocument.toJson()` method which does not include spaces. Sample output from driver:

> `{"food": "donut", "color": "green"}`

Because of this, I would rather not include spaces.

File: source/fundamentals/crud/compound-operations.txt [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+
+Find and Update One Document
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To find and update a document, use the ``findOneAndUpdate()`` method of the
+``MongoCollection`` class.
+
+The following example uses the ``findOneAndUpdate()`` method to find a
+document with the ``color`` field set to ``"green"`` and  update the
+``food`` field in that document to ``"pizza"``.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:67]: Thanks for the note. As monospace means a large number of things in our documentation (class name, method name, interface name, field name, database name, collection name...), I disagree with the statement that "monospace implies it's a field name".

In regards to the second point, I understood the resolution of the conversation being referenced as "both formats are acceptable until the documentation is complete and then a task will be to make all of them consistent." 

As I personally prefer the `\`\`"<string>"\`\`` syntax, and as I understand it both are currently acceptable, I would prefer to keep the document as is.

File: source/fundamentals/crud/compound-operations.txt [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+Avoiding A Race Condition
+-------------------------
+
+In this section, we will go through an example that shows how we can
+avoid a race condition with compound operations.
+
+For this example, let's imagine that we run a hotel with one room and that we
+have a small Java program to help us checkout this room to a guest.  
+
+We represent our one room with this document in MongoDB.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:181]:
```suggestion
The following document in MongoDB represents the room:
```

Thanks for the note. I agree that the above phrasing improves the page. I added "in MongoDB" as I hadn't established a MongoDB instance within the example yet.

File: source/fundamentals/crud/compound-operations.txt [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+The following example uses the ``findOneAndReplace()`` method to find a
+document with the ``color`` field set to ``"green"`` and  replace it
+with the following document:
+
+.. code-block:: json
+
+    {"music": "classical", "color": "green"}
+
+The example also uses a ``FindOneAndReplaceOptions`` instance to specify that
+the returned document should be in the state after our replace operation.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:128]: Thanks for the note. I agree this is confusing. I have added a note explaining that by default compound operations return your document in the state before the write operation.

Fix:

File: source/fundamentals/crud/compound-operations.txt (hunk 1 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:93]: Thanks for the note. I think the updates from https://github.com/mongodb/docs-java/pull/80#discussion_r632748711 (adding description of return value) solve this issue.

File: source/fundamentals/crud/compound-operations.txt [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+
+    {"_id": 1, "guest": null, "room": "Blue Room", "reserved": false}
+
+`The full code for this example is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperators.java>`__
+
+Example With Race Condition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's say our app uses this unsafe ``bookARoom`` method to checkout our room to
+a guest.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt]: Thanks for the note. I have removed any mention of "safe" or "unsafe" from the page.

Fix: https://github.com/mongodb/docs-java/pull/80/commits/9539287d341c355f47a3db3ac98a70150f0286e0 and https://github.com/mongodb/docs-java/pull/80/commits/58b9fa40678addbad3e864210ff68610305b8cab

File: source/fundamentals/crud/compound-operations.txt [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+For more information on the methods and classes in this section, see the
+following API documentation pages:
+
+- :java-docs:`findOneAndDelete() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson,com.mongodb.client.model.FindOneAndDeleteOptions)>`
+- :java-docs:`FindOneAndDeleteOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndDeleteOptions.html>`
+
+Avoiding A Race Condition
+-------------------------
+
+In this section, we will go through an example that shows how we can
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt]: Thanks for the note. I agree this sentence is incorrect and does not correctly describe the following content. 

Fix: https://github.com/mongodb/docs-java/pull/80/commits/4531d6a5fd721d179199167bf556c40899085d5e

File: source/fundamentals/crud/compound-operations.txt [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+
+The example also uses a ``FindOneAndUpdateOptions`` instance to specify the
+following options:
+
+- Exclude the ``_id`` field from the found document
+- Specify to insert a document if no documents match our query. MongoDB and
+  other databases refer to this operation as an upsert.
+- Set a maximum execution time of 5 seconds for this operation on the MongoDB
+  instance. If the operation takes longer, the ``findOneAndUpdate()`` method
+  will throw a ``MongoExecutionTimeoutException``.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:85]: After synchronous zoom discussion, it was agreed that this list is acceptable as is.

File: source/fundamentals/crud/compound-operations.txt [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+    {"_id": 2, "food": "pear", "color": "yellow"}
+
+`The full code for the following examples is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java>`__
+
+Find and Update One Document
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To find and update a document, use the ``findOneAndUpdate()`` method of the
+``MongoCollection`` class.
+
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:75]: Thanks for the note. This was discussed synchronously over zoom.

We agreed that this implementing this note would improve the flow of the page but would violate the "max depth 2 subheadings rule" in the [MongoDB Style Guide](https://docs.mongodb.com/meta/style-guide/style/titles-and-headings/#style-and-structure).

Fix: 

Plan to discuss this more with the team at a future team PR review meeting.

File: source/fundamentals/crud/compound-operations.txt [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+- Find and return an empty room for Jan
+- Find and return an empty room for Pat
+- Update the room to booked for Pat
+- Update the room to booked for Jan
+
+Notice that for a brief moment Pat had reserved the room, but as Jan's update
+operation was the last to execute our document has ``"Jan"`` as the guest. 
+
+Let's see how we can use compound operations to avoid this race condition and
+always give our users the correct message.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:258]: Thanks for the note. This note was discussed synchronously over zoom.

We agreed that the section of text above was not performing enough work in the text, and that it should be reworded in order to replace first sentence of the avoiding a race condition section. We agreed that the sentence above should be rephrased to be less referential to the "Example With Race Condition" section so that readers could skip immediately to this section of the text.

Fix:

File: source/fundamentals/crud/compound-operations.txt [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide you will learn how to perform **compound operations** with the
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:16]:
```suggestion
In this guide, you will learn how to perform **compound operations** with the
```

File: source/fundamentals/crud/compound-operations.txt [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+MongoDB Java driver. 
+
+Compound operations consist of a read and write operation performed as one
+**atomic operation**. An atomic operation is an operation which either completes
+entirely, or does not complete at all. Atomic operations cannot partially complete.
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:21]: Based on your current definition, it's implied that atomic operations cannot partially complete.

```suggestion
entirely, or does not complete at all.
```

File: source/fundamentals/crud/compound-operations.txt [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+
+Atomic operations can help you avoid **race conditions** in your code. A
+race condition occurs when your code's behavior is dependent on the order of
+uncontrollable events.
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:25]:
```suggestion
 events.
```
Uncontrollable seems wordy. Removing it still delivers the same message.

File: source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+    private void findOneAndUpdateExample() {
+        System.out.println("Starting find one and update example...");
+        MongoCollection<Document> collection = getCollection();
+        //start findOneAndUpdate-example
+        // <MongoCollection set up code here>
+        Bson projection = Projections.excludeId();
+        Bson filter = Filters.eq("color", "green");
+        Bson update = Updates.set("food", "pizza");
+        FindOneAndUpdateOptions options = new FindOneAndUpdateOptions().
```
Comment (kyuan-mongodb) [inline source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java:55]: I tried following this as part of my PR to use ``findOneAndUpdate``, but I had trouble retrieving the document after the update. By default, this returns the state of the document before the update. Similar to your ``findOneAndReplace`` example, you need to specify the ``returnDocument()``. (This may be the case for ``findOneAndDelete`` too)

```suggestion
        FindOneAndUpdateOptions options = new FindOneAndUpdateOptions().
        returnDocument(ReturnDocument.AFTER).
```

File: source/fundamentals/crud/compound-operations.txt [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+
+MongoDB supports the following compound operations:
+
+- Find and update one document
+- Find and replace one document
+- Find and delete one document
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:31]: This seems a bit out of place in the overview. I think this should go into your How to Use Compound Operations section and if you want, you can hyperlink each one to their respective sections.

File: source/fundamentals/crud/compound-operations.txt [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+For more information on atomic operations and atomicity, see 
+:manual:`the MongoDB manual entry for atomicity and transactions </core/write-operations-atomicity/>`.
+
+For more information on transactions, 
+:manual:`see the MongoDB manual entry for transactions </core/transactions/>`.
+
+How to Use Compound Operations
+------------------------------
+
+This section shows how to use each compound operation with the MongoDB Java Driver.
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:47]:
```suggestion

```

Suggestion: Replace this with the part in your intro about the compound operations in the java driver

File: source/fundamentals/crud/compound-operations.txt [hunk 25]
```diff
@@ -0,0 +1,313 @@
...
+`The full code for the following examples is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java>`__
+
+.. note:: Before or After the Write?
+
+   By default, compound operations return your found document in the state
+   before your write operation. You can retrieve your found document in the
+   state after your write operation by using the options class corresponding to
+   your compound operation. You can see an example of this configuration in the 
+   :ref:`Find and Replace example below <java-find-and-replace-example>`.
+
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:65]:
```suggestion
.. note:: Before or After the Write?

   By default, each compound operation returns your found document in the state
   before your write operation. You can retrieve your found document in the
   state after your write operation by using the options class corresponding to
   your compound operation. 

```

Comment (kyuan-mongodb) [reply source/fundamentals/crud/compound-operations.txt:65]: Suggestion: Mentioning that you return the state of the document in each example by appending ``returnDocument(ReturnDocument.AFTER)`` to the option makes your note make more sense.  At first, I was confused why you put it so high up.

Doing this would also avoid having to explain it in each section that this is what you're doing.

File: source/fundamentals/crud/compound-operations.txt [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+The following example uses the ``findOneAndUpdate()`` method to find a
+document with the ``color`` field set to ``"green"`` and  update the
+``food`` field in that document to ``"pizza"``. 
+
+The example also uses a ``FindOneAndUpdateOptions`` instance to specify the
+following options:
+
+- Exclude the ``_id`` field from the found document.
+- Specify to insert a document if no documents match the query. MongoDB and
+  other databases refer to this operation as an upsert.
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:85]:
```suggestion
- Specify an upsert, which inserts a document if no documents match the query.
```
Suggestion: It's more clear if you define the term after it is mentioned. As is, the user doesn't know what it's called till they get to the second sentense.

File: source/fundamentals/crud/compound-operations.txt [hunk 27]
```diff
@@ -0,0 +1,313 @@
...
+The output of the above code should look like this:
+
+.. code-block:: json
+
+   {"food": "donut", "color": "green"}
+
+For more information on the ``Projections`` class, see our
+:doc:`guide on the Projections builder </fundamentals/builders/projections/>`.
+
+For more information on the methods and classes in this section, see the
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:105]:
```suggestion
For more information on the method and classes in this section, see the
```
Nit: I noticed this in my own PR too, but "methods and classes" are in it's plural form when "methods" should be in singular.

Applies to all instances.

File: source/fundamentals/crud/compound-operations.txt [hunk 28]
```diff
@@ -0,0 +1,313 @@
...
+
+Avoiding a Race Condition
+-------------------------
+
+In this section we explore two examples. The first example contains a
+race condition, the second example uses a compound operation to
+avoid the race condition present in the first example.
+
+In this section, let's imagine that we run a hotel with one room and that we
+have a small Java program to help us checkout this room to a guest.
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:199]:
```suggestion
For both examples, we run a hotel with one room and use a Java program to help us checkout this room to a guest.  
```
Suggestion: Something along the lines of this wording to avoid starting 2 paragraphs with "in this section"

File: source/fundamentals/crud/compound-operations.txt [hunk 29]
```diff
@@ -0,0 +1,313 @@
...
+-------------------------
+
+In this section we explore two examples. The first example contains a
+race condition, the second example uses a compound operation to
+avoid the race condition present in the first example.
+
+In this section, let's imagine that we run a hotel with one room and that we
+have a small Java program to help us checkout this room to a guest.  
+
+The following document in MongoDB represents the room:
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:201]:
```suggestion
The following document represents the room:
```
Nit: I think MongoDB is implied here.

File: source/fundamentals/crud/compound-operations.txt [hunk 30]
```diff
@@ -0,0 +1,313 @@
...
+
+    {"_id": 1, "guest": null, "room": "Blue Room", "reserved": false}
+
+`The full code for this example is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperators.java>`__
+
+Example With Race Condition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's say our app uses this ``bookARoom`` method to checkout our room to
+a guest:
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:213]:
```suggestion
Our app uses this ``bookARoom`` method to checkout our room to
a guest:
```
Suggestion: Using "lets say" and "imagine" leaves too much room for imagination. Personally, when I read about an example, I expect concrete scenario's where I shouldn't be wondering about the opposite or tangent scenario.

Applies to all instances.

File: source/fundamentals/crud/compound-operations.txt [hunk 31]
```diff
@@ -0,0 +1,313 @@
...
+   Sorry, we are booked Pat
+
+When we look at our database, we see the following: 
+
+.. code-block:: json
+
+    {"_id":1, "guest":"Jan", "room":"Blue Room", "reserved":true}
+ 
+Pat got the correct message. While she might be sad she didn't get the
+reservation, at least she will know not to travel to our hotel.
```
Comment (kyuan-mongodb) [inline source/fundamentals/crud/compound-operations.txt:292]:
```suggestion
reservation, at least she knows not to travel to our hotel.
```
Nit: future tense

**Review by kyuan-mongodb — CHANGES_REQUESTED** [review]
Getting close! Most comments are wording suggestions and there are a few around how the methods work.

File: source/fundamentals/crud/compound-operations.txt (hunk 20 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:21]: Thanks for the note. As atomic operations and concurrency are a notoriously tricky concept, I would like to leave this in. I think it will benefit readers to have the corollary explicitly stated as the definition is dense, nuanced, and jargon heavy.

File: source/fundamentals/crud/compound-operations.txt (hunk 21 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:25]: Thanks for the note. I believe it is incorrect to say that this edit delivers the same meaning. All programs (and all happenings within a computer) are dependent on events. A person must start a program (an event), a program must move information within a machine (an event), etc... "Uncontrollable" is the main word in the definition of race condition here.

File: source/fundamentals/crud/compound-operations.txt (hunk 27 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:105]: Thanks for the note. I do not believe this edit benefits the page. As the linked method and classes will link to multiple methods and classes (the [`FindOneAndUpdateOptions`](https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndUpdateOptions.html) presents the reader with a list of methods 4 of which are used in the example), I feel changing "methods" to "method" will not improve accuracy (and is sound slightly off to my ear).

File: source/fundamentals/crud/compound-operations.txt (hunk 29 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:201]: As a document is also a class within the MongoDB Java Driver, I do not agree that MongoDB is implied here. This could be a document within the Driver, or a document in a BSON file stored in the filesystem alongside our application.

File: source/fundamentals/crud/compound-operations.txt (hunk 30 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:213]: Thanks for the note. As this example is explicitly designed to have you wondering about the opposite scenario (the opposite scenario is presented immediately after this example), so I feel it is appropriate to use open language like "let's say" and "imagine".

File: source/fundamentals/crud/compound-operations.txt (hunk 28 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:199]:
```suggestion
For both examples, let's imagine that we run a hotel with one room and that we 
have a small Java program to help us checkout this room to a guest.  
```

Good catch, I agree the double "in this section" is unproductive.

File: source/fundamentals/crud/compound-operations.txt (hunk 23 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:31]: I disagree, and feel that this information is necessary for a reader to determine if this is the page they are looking for. Without seeing the complete list of supported compound operators, I do not think a reader will be able to make a judgment call on whether they should read the compound ops page, the transactions page, or a different CRUD page.

Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:31]: This comment was discussed synchronously and it was agreed that this information is necessary for a reader to make a judgment call if they are on the correct page.

File: source/fundamentals/crud/compound-operations.txt (hunk 24 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:47]: Response to this edit is in the thread for https://github.com/mongodb/docs-java/pull/80#discussion_r634413543

File: source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java (hunk 22 above)
Comment (biniona-mongodb) [reply source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java:55]: This comment was discussed synchronously. We agreed this should not be updated to `returnDocumnet(ReturnDocument.AFTER)` as this may imply that the default behavior of the driver is undesirable if all examples do not use it. The solution arrived at was to add a comment above this line of code clarifying that this will return the document your update operation.

File: source/fundamentals/crud/compound-operations.txt (hunk 25 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:65]: This comment was discussed synchronously. It was agreed that the example link is necessary as only one example discusses this option. We agreed that the earlier edit around "each compound returns"... improved the page.

File: source/fundamentals/crud/compound-operations.txt (hunk 26 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/compound-operations.txt:85]: This comment was discussed synchronously. At first I did not agree with this edit, but I after discussion I agreed that this edit, paired with a link to a full description of upsert operations, improves the page.  

Fix:

**Review by kyuan-mongodb — APPROVED** [review]
LGTM!

**Comment by biniona-mongodb** [comment]
Hi @jyemin, could I please have a technical review for this PR?

**Review by jyemin — APPROVED** [review]
Nicely done!

LGTM
